ngrok/src/tunnel_ext.rs

```diff
@@ -0,0 +1,188 @@
+    ///
+    /// Provides slightly nicer errors when the backend is unavailable.
+    #[instrument(level = "debug", skip_all, fields(local_addrs))]
+    async fn forward_http(&mut self, addr: impl ToSocketAddrs + Send) -> Result<(), io::Error> {
+        forward_conns(self, addr, |e, c| drop(serve_gateway_error(e, c))).await
+    }
+}
+
+async fn forward_conns<T, A, F>(this: &mut T, addr: A, mut on_err: F) -> Result<(), io::Error>
+where
+    T: Tunnel + ?Sized,
+    A: ToSocketAddrs,
+    F: FnMut(io::Error, Conn),
+{
+    let span = Span::current();
+    let addrs = tokio::net::lookup_host(addr).await?.collect::<Vec<_>>();
+    span.record("local_addrs", field::debug(&addrs));
+    trace!("looked up local addrs");
+    loop {
+        trace!("waiting for new tunnel connection");
+        if !handle_one(this, addrs.as_slice(), &mut on_err).await? {
+            debug!("listener closed, exiting");
+            break;
+        }
+    }
+    Ok(())
+}
+
+async fn forward_bytes(
+    mut r: impl AsyncRead + Unpin,
+    mut w: impl AsyncWrite + Unpin,
+) -> Result<(), io::Error> {
+    let mut buf = vec![0u8; 1024];
+    loop {
+        let n = r.read(&mut buf).await?;
+        if n == 0 {
+            break;
+        }
+
+        w.write_all(&buf[0..n]).await?;
+    }
+    Ok(())
+}
+
+fn join_streams(
+    left: impl AsyncRead + AsyncWrite + Unpin + Send + 'static,
+    right: impl AsyncRead + AsyncWrite + Unpin + Send + 'static,
+) -> JoinHandle<()> {
+    let (t_rx, t_tx) = tokio::io::split(left);
+    let (l_rx, l_tx) = tokio::io::split(right);
+
+    let joined = futures::future::join(forward_bytes(t_rx, l_tx), forward_bytes(l_rx, t_tx));
+    tokio::spawn(
+        async move {
+            let (to_tunnel, to_local) = joined.await;
+            debug!(?to_tunnel, ?to_local, "connection closed");
+        }
+        .in_current_span(),
+    )
+}
+
+#[instrument(level = "debug", skip_all, fields(remote_addr, local_addr))]
+async fn handle_one<T, F>(
+    this: &mut T,
+    addrs: &[SocketAddr],
+    on_error: F,
+) -> Result<bool, io::Error>
+where
+    T: Tunnel + ?Sized,
+    F: FnOnce(io::Error, Conn),
+{
+    let span = Span::current();
+    let tunnel_conn = if let Some(conn) = this
+        .try_next()
+        .await
+        .map_err(|err| io::Error::new(io::ErrorKind::NotConnected, err))?
+    {
+        conn
+    } else {
+        return Ok(false);
+    };
+
+    span.record("remote_addr", field::debug(tunnel_conn.remote_addr()));
+
+    trace!("accepted tunnel connection");
+
+    let local_conn = match TcpStream::connect(addrs).await {
+        Ok(conn) => conn,
+        Err(error) => {
+            warn!(%error, "error establishing local connection");
+
+            on_error(error, tunnel_conn);
+
+            return Ok(true);
+        }
+    };
+    span.record("local_addr", field::debug(local_conn.peer_addr().unwrap()));
+
+    trace!("established local connection");
+
+    debug!("forwarding tunnel connection to local address");
```

Comment:
Will this debug not also be included at the trace level? Curious why there are both of these back to back.

Reply:
I think my original thought process was that they're two separate events - 1. Connected to the local address, 2. Starting the byte forwarding.

Can probably drop the "connected" one though, since it's kinda implied.